kafka/client.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kafka
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/kelseyhightower/envconfig"
+)
+
+type AdapterSASL struct {
+	Enable   bool   `envconfig:"KAFKA_NET_SASL_ENABLE" required:"false"`
+	User     string `envconfig:"KAFKA_NET_SASL_USER" required:"false"`
+	Password string `envconfig:"KAFKA_NET_SASL_PASSWORD" required:"false"`
+}
+
+type AdapterTLS struct {
+	Enable bool   `envconfig:"KAFKA_NET_TLS_ENABLE" required:"false"`
+	Cert   string `envconfig:"KAFKA_NET_TLS_CERT" required:"false"`
+	Key    string `envconfig:"KAFKA_NET_TLS_KEY" required:"false"`
+	CACert string `envconfig:"KAFKA_NET_TLS_CA_CERT" required:"false"`
+}
+
+type AdapterNet struct {
+	SASL AdapterSASL
+	TLS  AdapterTLS
+}
+
+type envConfig struct {
+	BootstrapServers []string `envconfig:"KAFKA_BOOTSTRAP_SERVERS" required:"true"`
+	Net              AdapterNet
+}
+
+// NewConfig extracts the Kafka configuration from the environment.
+func NewConfig(ctx context.Context) ([]string, *sarama.Config, error) {
+	cfg := sarama.NewConfig()
+	cfg.Version = sarama.V2_0_0_0
+
+	var env envConfig
+	if err := envconfig.Process("", &env); err != nil {
+		return nil, nil, err
+	}
+
+	if env.Net.SASL.Enable {
+		cfg.Net.SASL.Enable = true
+		cfg.Net.SASL.User = env.Net.SASL.User
+		cfg.Net.SASL.Password = env.Net.SASL.Password
+	}
+
+	if env.Net.TLS.Enable {
+		cfg.Net.TLS.Enable = true
+		tlsConfig, err := newTLSConfig(env.Net.TLS.Cert, env.Net.TLS.Key, env.Net.TLS.CACert)
+		if err != nil {
+			return nil, nil, err
+		}
+		cfg.Net.TLS.Config = tlsConfig
+	}
+
+	return env.BootstrapServers, cfg, nil
+}
+
+// NewConsumer is a helper method for constructing a client for consuming kafka messages.
+func NewConsumer(ctx context.Context) (sarama.Client, error) {
+	bs, cfg, err := NewConfig(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg.Consumer.Offsets.Initial = sarama.OffsetOldest
+	cfg.Consumer.Return.Errors = true
+
+	return sarama.NewClient(bs, cfg)
+}
+
+// NewProducer is a helper method for constructing a client for producing kafka methods.
+func NewProducer(ctx context.Context) (sarama.Client, error) {
```

Comment:
Since we are in the CloudEvents world, would it be interesting to have an additional method that returns a Cloudevents Client configured for Kafka?

Reply:
This is an example for a producer: https://github.com/cloudevents/sdk-go/blob/master/v2/cmd/samples/kafka/sender/main.go#L14

And for consumer: https://github.com/cloudevents/sdk-go/blob/master/v2/cmd/samples/kafka/receiver/main.go#L15

Reply:
This is all about making things simpler, but perhaps as a follow-up?

Reply:
Sure!